pyintegration/tests/test_dtypes.py

```diff
@@ -0,0 +1,194 @@
+            self.assertIn("[B", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertNotEqual(expected, py_array)
+
+    def test_floating_array_of(self):
+        with self.subTest("double array from numpy array with nan2value null2value"):
+            np_array = np.array([float('nan'), NULL_DOUBLE, 1.1, float('inf')], dtype=np.float64)
+            expected = [NULL_DOUBLE, 0, 1.1, float('inf')]
+            j_array = dtypes.float64.array_of(np_array, nan_to_value=NULL_DOUBLE, null_to_value=0.0)
+            self.assertIn("[D", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("double array from numpy array"):
+            np_array = np.array([float('nan'), NULL_DOUBLE, 1.1, float('inf')], dtype=np.float64)
+            pd_series = pd.Series(np_array)
+            j_array = dtypes.double.array_of(pd_series)
+            py_array = [x for x in j_array]
+            expected = [float('nan'), NULL_DOUBLE, 1.1, float('inf')]
+            self.assertTrue(math.isnan(py_array[0]))
+            self.assertEqual(expected[1:], py_array[1:])
+
+        with self.subTest("float array from numpy array, down cast from double to float"):
+            np_array = np.array([float('nan'), 1.1, float('inf')], dtype=np.float64)
+            expected = [NULL_FLOAT, 1.1, float('inf')]
+            j_array = dtypes.float32.array_of(np_array, nan_to_value=NULL_FLOAT)
+            self.assertIn("[F", str(type(j_array)))
+            py_array = [x for x in j_array]
+            for i in range(3):
+                self.assertAlmostEqual(expected[i], py_array[i])
+
+        with self.subTest("double array from numpy long array"):
+            expected = [NULL_LONG, 1, 2, 3]
+            np_array = np.array(expected, dtype=np.int64)
+            j_array = dtypes.float64.array_of(np_array)
+            self.assertIn("[D", str(type(j_array)))
+            py_array = [x for x in j_array]
+            for i in range(4):
+                self.assertAlmostEqual(expected[i], py_array[i])
+
+        with self.subTest("double array from Python list of integer"):
+            expected = [NULL_LONG, 1, 2, 3]
+            j_array = dtypes.float64.array_of(expected)
+            py_array = [x for x in j_array]
+            for i in range(3):
+                self.assertAlmostEqual(expected[i], py_array[i])
+
+    def test_char_array_of(self):
+        test_str = "abcdefg0123456"
+        j_array = dtypes.char.array_of(test_str)
+        self.assertIn("[C", str(type(j_array)))
+        py_array = [chr(x) for x in j_array]
+        self.assertEqual(test_str, "".join(py_array))
+
+    def test_datetime(self):
+        values = [dtypes.DateTime(round(time.time())), dtypes.DateTime.j_type.now()]
+        j_array = dtypes.DateTime.array_of(values)
+        self.assertTrue(2, len(j_array))
+        # TODO add content comparison
+
+
```

Comment:
Overall, there are a lot of tests here, but I have concerns.

1) The underlying code has many many branches, and the unit tests do not systematically test all of the input arguments.  I'm certain that our coverage is significantly less than we think.
2) Some of the unit tests seem to be written to match the values coming out of the code, instead of writing the tests with the desired values in mind.  e.g. short returning `0` instead of `null`, `Inf` getting mapped to `null`, etc.
3) Other functions are not tested, e.g. `to_char`.

Reply:
I think the remapping change should have the coverage very high.  Are we still missing anything in tests?